topgg/lib/webhooks.lua

```diff
@@ -0,0 +1,176 @@
+local http = require('coro-http')
+local json = require('json')
+local tls = require('tls')
+
+local Webhooks = {}
+
+Webhooks.__index = Webhooks
+
+function Webhooks:new_with_authorization(authorization)
+  local webhooks = setmetatable({}, self)
+
+  webhooks.authorization = authorization
+  webhooks.routes = {}
+
+  return webhooks
+end
+
+function Webhooks:new(authorization)
+  local webhooks = setmetatable({}, self)
+
+  webhooks.authorization = authorization
+  webhooks.routes = {}
+
+  return webhooks
+end
+
+function Webhooks:add(path, callback, authorization)
+  authorization = authorization or self.authorization
+
+  if type(callback) ~= 'function' then
+    error("argument 'callback' must be a function")
+  elseif type(authorization) ~= 'string' then
+    error('route must have a clear authorization')
+  end
+
+  table.insert(self.routes, {
+    path = path,
+    authorization = authorization,
+    callback = callback,
+  })
+end
+
+function Webhooks:start(input)
+  if not input or type(input.host) ~= 'string' or type(
+    input.port
+  ) ~= 'number' then
+    error('missing host and port')
+  end
+
+  if input.key and input.cert then
+    tls.createServer(
+      {
+        key = input.key,
+        cert = input.cert,
+      },
+      function(socket)
+        local body = ''
+
+        socket:on('data', function(chunk)
+          body = body .. chunk
+
+          if body:find('\r\n\r\n') then
```

Comment:
Why are we manually searching through a data payload? 😅 Is there really no better way in lua than a raw tls socket?